### PR TITLE
Add graph capture validation pass

### PR DESCRIPTION
**Pull request by derdeljanTT**
Implement a mlir pass which performs graph capture validation of output TTNN graph. By default, this pass is disabled. https://github.com/tenstorrent/tt-mlir/issues/1183

fyi @mbezuljTT @s-jovic

### CODE REVIEW

File: runtime/tools/python/ttrt/common/run.py [hunk 1]
```diff
@@ -450,6 +457,7 @@ def _execute(binaries):
                                     program_index,
                                     total_inputs[loop],
                                     total_outputs[loop],
+                                    self["--use-graph-capture"],
```
Comment (tapspatel) [inline runtime/tools/python/ttrt/common/run.py:460]: can we add this in trt.runtime.DebugEnv? It seems like this feature is a debug mode feature and the infra is already there to add specific bools like this one

debug_env = ttrt.runtime.DebugEnv.get(
                self["--load-kernels-from-disk"], self["--enable-async-ttnn"]
            )

File: runtime/lib/ttnn/program.cpp [hunk 2]
```diff
@@ -176,10 +220,25 @@ static bool handleNopProgram(::tt::target::ttnn::Program const *program,
   return isNop;
 }
 
+std::unique_ptr<ProgramExecutor>
+makeProgramExecutor(const TensorMap &liveTensors,
+                    const std::unordered_set<uint32_t> &programInputs,
+                    const std::unordered_set<uint32_t> &programOutputs,
+                    ::ttnn::MeshDevice *meshDevice, bool useGraphCapture) {
+  if (useGraphCapture) {
```
Comment (tapspatel) [inline runtime/lib/ttnn/program.cpp:228]: this would be something like debug::Env::get().useGraphCapture

**Review by tapspatel — CHANGES_REQUESTED** [review]
changes inlined

File: include/ttmlir/Dialect/TTNN/Pipelines/TTNNPipelines.h [hunk 3]
```diff
@@ -110,6 +110,14 @@ struct TTIRToTTNNBackendPipelineOptions
 
   ListOption<int64_t> meshShape{
       *this, "mesh-shape", llvm::cl::desc("Set the multi-device mesh shape.")};
+
+  // If this option is true, run the entire graph with graph capture to validate
+  // it.
+  //
+  Option<bool> graphCaptureValidationEnabled{
```
Comment (nobradovictt) [inline include/ttmlir/Dialect/TTNN/Pipelines/TTNNPipelines.h:117]: You lack a test using this option.

File: lib/Dialect/TTNN/Transforms/Passes.cpp [hunk 4]
```diff
@@ -98,6 +111,66 @@ class TTNNDeallocate : public impl::TTNNDeallocateBase<TTNNDeallocate> {
...
+    runGraphCapture(flatBufferFilePath, outReportPath);
+
+    if (!isValidGraphCaptureReport(outReportPath)) {
+      // TODO (nobradovic/odjuricic): Handle recompile.
+    }
+  }
+
+  void outputTTNNIRFile(const std::string &mlirFilePath) {
+    ModuleOp module = getOperation();
+    std::error_code _ec;
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Transforms/Passes.cpp:140]: What happens in case of failure?

File: lib/Dialect/TTNN/Transforms/Passes.cpp [hunk 5]
```diff
@@ -98,6 +111,66 @@ class TTNNDeallocate : public impl::TTNNDeallocateBase<TTNNDeallocate> {
...
+  void runOnOperation() final {
+    const std::filesystem::path tmpDirPath =
+        std::filesystem::temp_directory_path();
+
+    const std::string mlirFilePath = tmpDirPath / "module.mlir";
+    const std::string flatBufferFilePath = tmpDirPath / "module.ttnn";
+    const std::string outReportPath = tmpDirPath / "module_graph_capture.json";
+
+    outputTTNNIRFile(mlirFilePath);
+    outputFlatBufferFile(mlirFilePath, flatBufferFilePath);
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Transforms/Passes.cpp:130]: How are failures in each stage handled, how do you know you are not reading artefact of some previous compile session?

File: lib/Dialect/TTNN/Transforms/Passes.cpp [hunk 6]
```diff
@@ -98,6 +111,66 @@ class TTNNDeallocate : public impl::TTNNDeallocateBase<TTNNDeallocate> {
...
+        " -o " + flatBufferFilePath;
+
+    system(cmd.c_str());
+  }
+
+  void runGraphCapture(const std::string &flatBufferFilePath,
+                       const std::string &outReportFilePath) {
+    // TODO(mbezulj): Add required env variable to be able to run graph capture
+    // with mockup device and without kernel compilation.
+    const std::string cmd = "ttrt run " + flatBufferFilePath +
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Transforms/Passes.cpp:158]: Do we have something like std::format to allow plugin of parameters within const string?

File: lib/Dialect/TTNN/Transforms/Passes.cpp [hunk 7]
```diff
@@ -98,6 +111,66 @@ class TTNNDeallocate : public impl::TTNNDeallocateBase<TTNNDeallocate> {
...
+
+    const std::string mlirFilePath = tmpDirPath / "module.mlir";
+    const std::string flatBufferFilePath = tmpDirPath / "module.ttnn";
+    const std::string outReportPath = tmpDirPath / "module_graph_capture.json";
+
+    outputTTNNIRFile(mlirFilePath);
+    outputFlatBufferFile(mlirFilePath, flatBufferFilePath);
+    runGraphCapture(flatBufferFilePath, outReportPath);
+
+    if (!isValidGraphCaptureReport(outReportPath)) {
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Transforms/Passes.cpp:133]: IsValid needs to return more than bool. It needs to be loc<->op type mapping to ->  exception type.

Comment (nobradovictt) [reply lib/Dialect/TTNN/Transforms/Passes.cpp:133]: Then handlers need to be written for every exception type.

File: runtime/lib/ttnn/program.cpp [hunk 8]
```diff
@@ -51,12 +57,49 @@ class ProgramExecutor {
...
+  void execute(const ::tt::target::ttnn::Program *program) override {
+    const auto execute_impl = [&]() {
+      unsigned int opIndex = 0;
+      for (const ::tt::target::ttnn::Operation *op : *program->operations()) {
+        LOG_DEBUG(LogType::LogRuntimeTTNN,
+                  "Executing operation: ", op->debug_info()->c_str());
+
+        try {
+          runOperation(op);
+        } catch (const std::exception &ex) {
```
Comment (nobradovictt) [inline runtime/lib/ttnn/program.cpp:79]: Can this leave device in a bad state? Can it leak memory?

**Comment by mbezuljTT** [comment]
Thanks everyone for great feedback on the implementation! I'll close PR per explanation I've wrote on the issue #1183
I think we have a good plan to move forward, but it will take us some time to get there.

**Pull request closed without merging**